certbot-apache/certbot_apache/interfaces.py

```diff
@@ -123,10 +123,19 @@ class ParserNode(object):
     dirty: bool
 
     # Filepath of the file where the configuration element for this ParserNode
-    # object resides.
+    # object resides. This can be None if a configuration directive is defined in
+    # for example the httpd command line.
     filepath: Optional[str]
     """
 
+    @abc.abstractmethod
+    def __init__(self, ancestor=None, filepath="", dirty=False):  # pragma: no cover
+        """
+        Initializes the ParserNode instance, and sets the ParserNode specific
+        instance variables. This is not meant to be used directly, but through
+        specific classes implementing ParserNode interface.
```

Comment:
If we take the approach I described at https://github.com/certbot/certbot/pull/7330#issuecomment-522795095, we should be sure to document the parameters and their defaults for each class in their docstring.

Reply:
Oh wow, indeed. Sorry about this.